### PR TITLE
first commit for connecting-to-databases.md and renamed visualizing-s…

### PR DESCRIPTION
**Pull request by TegveerG**
This contributes to #5

### CODE REVIEW

File: course-material/intro-to-sql/connecting-to-databases.md [hunk 1]
```diff
@@ -1,14 +1,606 @@
...
+    notebook_metadata_filter: myst
     text_representation:
       extension: .md
       format_name: markdown
       format_version: '1.3'
-      jupytext_version: 1.14.5
+      jupytext_version: 1.14.6
+  kernelspec:
+    display_name: Python 3 (ipykernel)
+    language: python
```
Comment (lfunderburk) [inline course-material/intro-to-sql/connecting-to-databases.md:13]: Please retain the set up below. I built the docs locally and it works ok. 

```
---
jupyter:
  jupytext:
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.14.5
---
```

File: course-material/intro-to-sql/connecting-to-databases.md [hunk 2]
```diff
@@ -1,14 +1,606 @@
...
+)
+```
+
+You're all set
+
+```{code-cell} ipython3
+%sql select * from penguins.csv limit 3
+```
+
+Example: show the first 5 rows.
```
Comment (lfunderburk) [inline course-material/intro-to-sql/connecting-to-databases.md:310]: The following section feels disjointed and would be hard to follow as the user is interacting with the penguins dataset. 

Suggestion: Remove the following as it is already covered in the EDA example
```
Example: show the first 5 rows.

```sql
SELECT *
FROM absenteeism 
LIMIT 5
```

#### Question 1.1 (Easy):
How many records are there in the 'absenteeism' table? 



<details>

<summary>Answers</summary>

You can use the `%%sql` magic and the `COUNT(*)` function to count the total number of records. 

```python
%%sql
SELECT COUNT(*) 
FROM absenteeism
```
</details>


#### Question 1.2 (Easy):
How many unique employees are listed in the dataset?




<details>

<summary>Answers</summary>

You can use the `%%sql` magic and the `COUNT(DISTINCT ID)` function to count the total number of unique instances of the `Age` column. 

```python
%%sql
SELECT COUNT(DISTINCT ID) 
FROM absenteeism;
```
</details>


#### Question 1.3 (Easy):
What is the average distance from residence to work? 



<details>

<summary>Show Answers</summary>

You can use the `%%sql` magic and the `AVG(Distance_from_Residence_to_Work)` function to calculate the average distance from residence to work.. 

```python
%%sql
SELECT AVG(Distance_from_Residence_to_Work) 
FROM absenteeism;
```
</details>


#### Question 2.1 (Medium):
On which days of the week does the average absenteeism time exceed 4 hours? 



<details>

<summary>Answers</summary>

You can use the `%%sql` magic and break down the query as follows:

1. Select the column with name `Day_of_the_week`
2. From the table called `absenteeism`
3. Then group the values by day of the week that have an average value (use `AVG`) of more than 4 hours in absenteeism. 

```python
%%sql
SELECT Day_of_the_week 
FROM absenteeism 
GROUP BY Day_of_the_week 
HAVING AVG(Absenteeism_time_in_hours) > 4;
```
</details>


#### Question 2.2 (Medium):
What is the average transportation expense for each season?



<details>

<summary>Answers</summary>

You can use the `%%sql` magic and. Use the `AVG(Transportation_expense)` with the alias `AVG_Transportation_Expense` function to count the average transporation expense, then group by seasons.

```python
%%sql
SELECT Seasons, AVG(Transportation_expense) AS AVG_Transportation_Expense
FROM absenteeism 
GROUP BY Seasons;
```
</details>



#### Question 2.3 (Medium):


What is the average absenteeism time for employees with BMI higher than the average BMI




<details>

<summary>Show Answers</summary>

You can use the `%%sql` magic and. Use the `AVG(Absenteeism_time_in_hours)` with the alias `AVG_Absenteeism_time_in_hours` function to count the average absenteeism (time units hours). 

`WHERE Body_mass_index > (`: This part begins a condition that the data must meet to be included in our average calculation. Here, we're only interested in rows where the `Body_mass_index` is greater than a certain value.

`SELECT AVG(Body_mass_index) FROM absenteeism)`: This is a subquery, a query within a query. It's calculating the average `Body_mass_index` for the entire absenteeism table.

```python
%%sql
SELECT AVG(Absenteeism_time_in_hours) as AVG_Absenteeism_time_in_hours
FROM absenteeism 
WHERE Body_mass_index > (
    SELECT AVG(Body_mass_index) 
    FROM absenteeism);
```
</details>


#### Question 3.1 (Hard):
Find the top 3 ages with the highest total absenteeism hours, excluding disciplinary failures.






<details>

<summary>Answers</summary>

You can use the `%%sql` magic and break down the query as follows:

1. Select the column with name `Age`, compute the Sum of `Absenteeism_time_in_hours`. Give this sum an alias `Sum_Absenteeism`.
2. From the table called `absenteeism`
3. The keywork WHERE is used to filter the data that meets a specific condition, in this case `Disciplinary_failure` is equal to zero.
4. Group values by the `Age` column.
5. Sort the values by the sum and show the first 3 values.

```python
%%sql
SELECT Age, SUM(Absenteeism_time_in_hours) AS Sum_Absenteeism
FROM absenteeism 
WHERE Disciplinary_failure = 0 
GROUP BY Age 
ORDER BY Sum_Absenteeism
DESC LIMIT 3;
```
</details>


#### Question 3.2 (Hard):

Find the age of employees who have been absent for more than 5 hours with an unjustified absence.

Hint: investigate encoding on the data source.



<details>

<summary>Answers</summary>

You can use the `%%sql` magic. 'Unjustified absence' is coded with 26. From there all that is required is selecting the age, and using `WHERE` to set up the appropriate conditions. 

```python
%%sql
SELECT Age 
FROM absenteeism 
WHERE Reason_for_absence = 26 AND Absenteeism_time_in_hours > 5;
```
</details>


#### Question 3.3 (Hard):

Which reasons for absence are more frequent for social drinkers than social non-drinkers?



<details>

<summary>Show Answers</summary>

You can use the `%%sql` magic. We use `SELECT` to extract the `Reason_for_absence` from the `absenteeism` table. 

The column `Social_drinker` is encoded using binary notation, 0=is not a social drinker, 1=is a social drinker. 

We next group by their reason for absence. 

`HAVING COUNT() > (`  begins the condition that the groups must meet to be included in the results. Only groups where the count of rows (representing the number of instances of each `Reason_for_absence` among social drinkers) is greater than a certain value will be included.

`SELECT COUNT() FROM absenteeism WHERE Social_drinker = 0 GROUP BY Reason_for_absence)`  is a subquery that calculates the count of rows for each `Reason_for_absence` where `Social_drinker` is 0 (indicating the employee is not a social drinker), effectively giving us the number of instances of each `Reason_for_absence` among non-social drinkers.

```python
%%sql
SELECT Reason_for_absence 
FROM absenteeism 
WHERE Social_drinker = 1 
GROUP BY Reason_for_absence 
HAVING COUNT() > (
    SELECT COUNT() 
    FROM absenteeism 
    WHERE Social_drinker = 0 
    GROUP BY Reason_for_absence);
```
</details>


## Bonus
```

File: course-material/intro-to-sql/connecting-to-databases.md [hunk 3]
```diff
@@ -1,14 +1,606 @@
...
+- [ClickHouse](https://jupysql.ploomber.io/en/latest/integrations/clickhouse.html)
+- [MariaDB](https://jupysql.ploomber.io/en/latest/integrations/mariadb.html)
+- [MindsDB](https://jupysql.ploomber.io/en/latest/integrations/mindsdb.html)
+- [MSSQL](https://jupysql.ploomber.io/en/latest/integrations/mssql.html)
+- [MySQL](https://jupysql.ploomber.io/en/latest/integrations/mysql.html)
+- [QuestDB](https://jupysql.ploomber.io/en/latest/integrations/questdb.html)
+- [Oracle](https://jupysql.ploomber.io/en/latest/integrations/oracle.html)
+- [Trino](https://jupysql.ploomber.io/en/latest/integrations/trinodb.html)
+
++++
```
Comment (lfunderburk) [inline course-material/intro-to-sql/connecting-to-databases.md:602]: Suggestion: add one to three hands on questions that allow people to practice their skills. 

For example:

1. Form a connection string for a postgres database
2. Form a connection string for a DuckDB instance
3. What is the difference between using MySQL and PostgreSQL
4. Load a csv into a DuckDB instance

The questions above are suggestions. The goal is to help them practice via the Google colab. You can use a similar format as the absenteeism dataset where you have an open ended question, you leave a cell open for their answer, and using markdown you can hide the answer.

**Review by lfunderburk — CHANGES_REQUESTED** [review]
This is a great start @TegveerG!

Added a few suggestions. In general the comments revolve around creating more relevant exercises. 

Please revert the config info at the top of the markdown file

```
---
jupyter:
  jupytext:
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.14.5
---
```

File: course-material/intro-to-sql/connecting-to-databases.md (hunk 2 above)
Comment (TegveerG) [reply course-material/intro-to-sql/connecting-to-databases.md:310]: Agreed, I added these questions as a placeholder while I was drafting my questions, shall include the new questions by today and commit

File: course-material/intro-to-sql/connecting-to-databases.md (hunk 3 above)
Comment (TegveerG) [reply course-material/intro-to-sql/connecting-to-databases.md:602]: Understood, thanks for the detailed feedback, incorporating them now and committing soon.

**Comment by TegveerG** [comment]
Another question I had was whether we should we run `pkgmt format` and `pkgmt lint` for these docs (I dont think it's currently present in the conda env)?

**Comment by lfunderburk** [comment]
This is a good question.
@edublancas what would you advice in this case?

**Comment by edublancas** [comment]
@TegveerG, @lfunderburk yes, we need to add linting. I added the config so `pkgmt lint` runs on each PR: https://github.com/ploomber/sql/issues/13 - there are some fixes required

**Comment by TegveerG** [comment]
Hi @lfunderburk, I get linting errors for Jonathan's and EDA files. I am still getting the `WARNING: Found an unexpected 'code-cell' or 'raw-cell' directive.` message when building and cannot figure whether my code is running fine.

**Review by lfunderburk — DISMISSED** [review]
This looks great! Thank you for adding the question and answer. These are good questions,

**Comment by TegveerG** [comment]
@edublancas I can't seem to fix the `E402 module level import not at top of file` linting errors in the `connecting-to-databases.md` file.

**Comment by edublancas** [comment]
you are adding noqa (which will solve the issue) to some imports:

image

but you still have some that don't have noqa:

image

I'd recommend you move all the imports to the top cell to get rid of this, it's a cleaner solution than using noqa - noqa is still helpful in other situations

**Review by lfunderburk — APPROVED** [review]
Content is looking finalized and ready. Waiting for the linting issue fix to be pushed based on Eduardo's suggestion

**Comment by TegveerG** [comment]
@lfunderburk fixed linting, imports are at the top now. I did find this a bit counterintuitive because when I was working with the `doc/api/magic-profile.md` file, I had imports throughout the markdown but no linting errors then.

**Review by lfunderburk — APPROVED** [review]
Ready for merge
